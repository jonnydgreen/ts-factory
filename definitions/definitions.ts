import { BlockInput } from './block/block.type.ts';
import { FunctionDeclarationInput } from './function-declaration/function-declaration.type.ts';
import { IdentifierInput } from './identifier/identifier.type.ts';
import { KeywordTypeNodeInput } from './keyword-type-node/keyword-type-node.type.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { SourceFileInput } from './source-file/source-file.type.ts';

// Definition
export type Definition =
  | SourceFileInput
  | StatementInput
  | ExpressionInput
  | TypeNodeInput
  | ModifierInput;

// Member Name
export type MemberNameInput = IdentifierInput;

// Statement Inputs
export type StatementInput = FunctionDeclarationInput | BlockInput;

// Expression Inputs
export type ExpressionInput = IdentifierInput;

// Type Node Inputs
export type TypeNodeInput = KeywordTypeNodeInput;

// Input
export type Input = SourceFileInput;
