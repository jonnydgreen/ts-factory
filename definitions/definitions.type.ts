import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type {
  ExpressionInput,
  MemberNameInput,
  PropertyNameInput,
  StatementInput,
  TypeElementInput,
  TypeNodeInput,
} from './definitions.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { TokenInput } from './token/token.type.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  T extends string | ts.PropertyName ? string | PropertyNameInput :
  T extends ts.Modifier ? ModifierInput :
  T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.TypeElement ? TypeElementInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  T extends ts.Token<ts.SyntaxKind> ? TokenInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;

export type PickDefinitionFields<T> = Omit<
  T,
  '__instructions' | 'leadingTrivia' | 'kind'
>;

export type DefinitionFields<T> = keyof PickDefinitionFields<T>;
