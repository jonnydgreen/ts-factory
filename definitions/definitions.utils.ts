import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type { ExpressionInput, MemberNameInput, StatementInput } from './definitions.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  // T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;
