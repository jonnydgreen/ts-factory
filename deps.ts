// Path
import { join, resolve } from 'https://deno.land/std@0.147.0/path/mod.ts';
export const Path = {
  join,
  resolve,
};

// TS
export { ts } from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
export * as TSM from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
