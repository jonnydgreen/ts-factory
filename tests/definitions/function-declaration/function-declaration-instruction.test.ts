import { Definition } from '../../../definitions/definitions.ts';
import { FunctionDeclarationInput } from '../../../definitions/function-declaration/function-declaration.type.ts';
import { ts, tsm } from '../../../deps.ts';
import {
  generateInstructions,
  processInstructions,
} from '../../../instructions/instructions.ts';
import { InstructionType } from '../../../instructions/instructions.type.ts';
import {
  assertIsError,
  assertSnapshot,
  assertThrows,
  blocks,
} from '../../../test.deps.ts';
import { assertTSMNodeKind } from '../../../utils/utils.assert.ts';
import {
  createSourceFile,
  createTestName,
  TestDefinition,
  TestDefinitionFields,
} from '../../test-utils.ts';

blocks.describe('Function Declaration', () => {
  blocks.describe('definition', () => {
    const definitions: TestDefinition<Definition>[] = [
      {
        name: createTestName('should process a Function Declaration'),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [{
            kind: ts.SyntaxKind.FunctionDeclaration,
            name: 'hello',
            parameters: [],
            body: { kind: ts.SyntaxKind.Block, statements: [] },
          }],
        },
      },
      {
        name: createTestName(
          'should process a Function Declaration',
          'with a object name definition',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [{
            kind: ts.SyntaxKind.FunctionDeclaration,
            name: {
              kind: ts.SyntaxKind.Identifier,
              text: 'hello',
            },
            parameters: [],
            body: { kind: ts.SyntaxKind.Block, statements: [] },
          }],
        },
      },
      {
        name: createTestName(
          'should process a Function Declaration',
          'with modifiers',
        ),
        input: {
          kind: ts.SyntaxKind.SourceFile,
          statements: [{
            kind: ts.SyntaxKind.FunctionDeclaration,
            name: 'hello',
            parameters: [],
            body: { kind: ts.SyntaxKind.Block, statements: [] },
            modifiers: [
              { kind: ts.SyntaxKind.ExportKeyword },
              { kind: ts.SyntaxKind.AsyncKeyword },
            ],
          }],
        },
      },
    ];
    for (const definition of definitions) {
      blocks.it({
        name: definition.name,
        fn: async (t) => {
          // Arrange
          const node = createSourceFile(definition.sourceFileContents);
          const instructions = generateInstructions(node, definition.input);

          // Act
          processInstructions(node, instructions);

          // Assert
          await assertSnapshot(t, node.getFullText());
        },
        ignore: definition.ignore,
        only: definition.only,
      });
    }
  });

  blocks.describe('fields', () => {
    const fieldDefinitions: TestDefinitionFields<FunctionDeclarationInput> = {
      type: [
        {
          name: createTestName('should SET the return type of a Function Declaration'),
          input: {
            kind: ts.SyntaxKind.FunctionDeclaration,
            parameters: [],
            type: {
              kind: ts.SyntaxKind.VoidKeyword,
            },
          },
          sourceFileContents: 'export function hello() {}',
        },
        {
          name: createTestName('should UNSET the return type of a Function Declaration'),
          input: {
            kind: ts.SyntaxKind.FunctionDeclaration,
            __instructions: {
              rules: [{
                instruction: InstructionType.UNSET,
                condition: '$exists(type)',
                field: 'type',
              }],
            },
            parameters: [],
          },
          sourceFileContents: 'export function hello(): void {}',
        },
      ],
      modifiers: [
        {
          name: createTestName('should ADD modifiers to the Function Declaration'),
          input: {
            kind: ts.SyntaxKind.FunctionDeclaration,
            parameters: [],
            modifiers: [
              { kind: ts.SyntaxKind.ExportKeyword },
              { kind: ts.SyntaxKind.AsyncKeyword },
            ],
          },
          sourceFileContents: 'function hello() {}',
        },
        {
          name: createTestName(
            'should ADD modifiers to the Function Declaration with existing modifiers',
          ),
          input: {
            kind: ts.SyntaxKind.FunctionDeclaration,
            parameters: [],
            modifiers: [
              { kind: ts.SyntaxKind.AsyncKeyword },
              { kind: ts.SyntaxKind.DefaultKeyword },
            ],
          },
          sourceFileContents: 'export async function hello() {}',
        },
        {
          name: createTestName(
            'should throw an error if an unsupported modifier is defined',
          ),
          input: {
            kind: ts.SyntaxKind.FunctionDeclaration,
            parameters: [],
            modifiers: [
              { kind: ts.SyntaxKind.ReadonlyKeyword },
            ],
          },
          sourceFileContents: 'function hello() {}',
          error: {
            prototype: TypeError,
            message: 'Unsupported Modifier kind ReadonlyKeyword',
          },
        },
      ],
    };
    for (const [field, definitions] of Object.entries(fieldDefinitions)) {
      blocks.describe(field, () => {
        for (const definition of definitions) {
          blocks.it({
            name: definition.name,
            fn: async (t) => {
              // Arrange
              const sourceFile = createSourceFile(definition.sourceFileContents);
              const [node] = getNodesOfKind(
                sourceFile,
                ts.SyntaxKind.FunctionDeclaration,
              );
              assertTSMNodeKind(node, ts.SyntaxKind.FunctionDeclaration);
              const instructions = generateInstructions(node, definition.input);

              if (definition.error) {
                // Act
                const result = assertThrows(() =>
                  processInstructions(node, instructions)
                );

                // Assert
                assertIsError(
                  result,
                  definition.error.prototype,
                  definition.error.message,
                );
              } else {
                // Act
                processInstructions(node, instructions);

                // Assert
                await assertSnapshot(t, node.getFullText());
              }
            },
            ignore: definition.ignore,
            only: definition.only,
          });
        }
      });
    }
  });
});

// What do we want to test?
//  - Y - Adding a new Definition
//  - Y - Adding a new Definition with different options
//  - Manipulating fields on the definition

// TODO: move
function getNodesOfKind<
  TKind extends ts.SyntaxKind,
  TNode extends tsm.KindToNodeMappings[TKind],
>(currentNode: tsm.Node, kind: TKind): TNode[] {
  const nodes: TNode[] = [];
  if (currentNode.isKind(kind)) {
    nodes.push(currentNode as TNode);
  }

  currentNode.forEachChild((childNode) => {
    nodes.push(...getNodesOfKind(childNode, kind) as TNode[]);
  });
  return nodes;
}
