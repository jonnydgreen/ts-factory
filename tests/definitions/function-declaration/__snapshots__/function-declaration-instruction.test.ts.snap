export const snapshot = {};

snapshot[`Function Declaration > definition > should process a Function Declaration 1`] = `
"function hello() { }

"
`;

snapshot[`Function Declaration > definition > should process a Function Declaration with a object name definition 1`] = `
"function hello() { }

"
`;

snapshot[`Function Declaration > definition > should process a Function Declaration with modifiers 1`] = `
"export async function hello() { }

"
`;

snapshot[`Function Declaration > fields > type > should SET the return type of a Function Declaration 1`] = `"export function hello(): void { }"`;

snapshot[`Function Declaration > fields > type > should UNSET the return type of a Function Declaration 1`] = `"export function hello() { }"`;

snapshot[`Function Declaration > fields > modifiers > should ADD modifiers to the Function Declaration 1`] = `"export async function hello() { }"`;

snapshot[`Function Declaration > fields > modifiers > should ADD modifiers to the Function Declaration with existing modifiers 1`] = `"export default async function hello() { }"`;
